File: flipperviewlib/src/main/java/com/juniperphoton/flipperviewlib/animation/MtxRotationAnimation.java
package com.juniperphoton.flipperviewlib.animation;

import android.graphics.Camera;
import android.graphics.Matrix;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.Transformation;

public class MtxRotationAnimation extends Animation {
    public final static int ROTATION_X = 1;
    public final static int ROTATION_Y = 1 << 1;

    private int mCenterX, mCenterY;
    private Camera mCamera = new Camera();

    private int mFromDeg;
    private int mToDeg;
    private int mRotationAxis = ROTATION_X;

    private boolean mUsePerspective = true;

    public MtxRotationAnimation(int rotationAxis, int fromDeg, int toDeg, int duration) {
        mFromDeg = fromDeg;
        mToDeg = toDeg;
        mRotationAxis = rotationAxis;
        setDuration(duration);
    }

    public void setUsePerspective(boolean usePerspective) {
        mUsePerspective = usePerspective;
    }

    @Override
    public void initialize(int width, int height, int parentWidth, int parentHeight) {
        super.initialize(width, height, parentWidth, parentHeight);

        mCenterX = width / 2;
        mCenterY = height / 2;

        setFillAfter(true);
    }

    @Override
    protected void applyTransformation(float interpolatedTime, Transformation transform) {
        super.applyTransformation(interpolatedTime, transform);

        int targetDeg = (int) (mFromDeg + (mToDeg - mFromDeg) * interpolatedTime);

        Matrix matrix = transform.getMatrix();
        mCamera.save();
        if ((mRotationAxis & ROTATION_X) == ROTATION_X) {
            mCamera.rotateX(targetDeg);
        }
        if ((mRotationAxis & ROTATION_Y) == ROTATION_Y) {
            mCamera.rotateY(targetDeg);
        }
        mCamera.getMatrix(matrix);
        matrix.preTranslate(-mCenterX, -mCenterY);
        matrix.postTranslate(mCenterX, mCenterY);
        mCamera.restore();
    }

    public static void setRotationX(View view, int deg) {
        MtxRotationAnimation animation = new MtxRotationAnimation(ROTATION_X, deg, deg, 1);
        view.startAnimation(animation);
    }

    public static void setRotationY(View view, int deg) {
        MtxRotationAnimation animation = new MtxRotationAnimation(ROTATION_Y, deg, deg, 1);
        view.startAnimation(animation);
    }
}

File: flipperviewlib/src/main/java/com/juniperphoton/flipperviewlib/animation/MtxRotationAnimation.kt
package com.juniperphoton.flipperviewlib.animation

import android.graphics.Camera
import android.view.View
import android.view.animation.Animation
import android.view.animation.Transformation

class MtxRotationAnimation(axis: Int, from: Int, to: Int, duration: Long) : Animation() {
    companion object {
        val ROTATION_X = 1
        val ROTATION_Y = 1 shl 1

        fun setRotation(rotationFun: () -> Int, view: View, deg: Int) {
            val animation = MtxRotationAnimation(rotationFun(), deg, deg, 1)
            view.startAnimation(animation)
        }
    }

    var centerX = 0
    var centerY = 0

    val camera = Camera()

    var fromDeg = 0
    var toDeg = 0
    var rotationAxis = ROTATION_X

    init {
        rotationAxis = axis
        fromDeg = from
        toDeg = to
        setDuration(duration)
    }

    override fun initialize(width: Int, height: Int, parentWidth: Int, parentHeight: Int) {
        super.initialize(width, height, parentWidth, parentHeight)

        centerX = width / 2
        centerY = height / 2

        fillAfter = true
    }

    override fun applyTransformation(interpolatedTime: Float, transform: Transformation?) {
        super.applyTransformation(interpolatedTime, transform)

        val targetDeg = (fromDeg + (toDeg - fromDeg) * interpolatedTime).toInt()

        val matrix = transform!!.matrix
        camera.save()
        if (rotationAxis and ROTATION_X == ROTATION_X) {
            camera.rotateX(targetDeg.toFloat())
        }
        if (rotationAxis and ROTATION_Y == ROTATION_Y) {
            camera.rotateY(targetDeg.toFloat())
        }
        camera.getMatrix(matrix)
        matrix.preTranslate((-centerX).toFloat(), (-centerY).toFloat())
        matrix.postTranslate(centerX.toFloat(), centerY.toFloat())
        camera.restore()
    }
}